src/fairchem/core/modules/element_references.py

```diff
@@ -0,0 +1,206 @@
+
+    @torch.autocast(device_type="cuda", enabled=False)
+    def dereference(
+        self, target: torch.Tensor, batch: Batch, reshaped: bool = True
+    ) -> torch.Tensor:
+        """Remove linear references"""
+        return self._apply_refs(target, batch, -1, reshaped=reshaped)
+
+    @torch.autocast(device_type="cuda", enabled=False)
+    def forward(
+        self, target: torch.Tensor, batch: Batch, reshaped: bool = True
+    ) -> torch.Tensor:
+        """Add linear references"""
+        return self._apply_refs(target, batch, 1, reshaped=reshaped)
+
+
+def create_element_references(
+    file: str | Path | None = None,
+    state_dict: dict | None = None,
+) -> LinearReference:
+    """Create an element reference module.
+
+    Currently only linear references are supported.
+
+    Args:
+        type (str): type of reference (only linear implemented)
+        file (str or Path): path to pt or npz file
+        state_dict (dict): a state dict of a element reference module
+
+    Returns:
+        LinearReference
+    """
+
+    # path takes priority if given
+    if file is not None:
+        try:
+            # try to load a pt file
+            state_dict = torch.load(file)
+        except RuntimeError:  # try to read an npz file
+            state_dict = {}
+            with np.load(file) as values:
+                # legacy linref files:
+                if "coeff" in values:
+                    state_dict["element_references"] = torch.tensor(values["coeff"])
+                else:
+                    state_dict["element_references"] = torch.tensor(
+                        values["element_references"]
+                    )
+
+    if "element_references" not in state_dict:
+        raise RuntimeError("Unable to load linear element references!")
+
+    return LinearReference(element_references=state_dict["element_references"])
+
+
+def fit_linear_references(
+    targets: list[str],
+    dataset: Dataset,
+    batch_size: int,
+    num_batches: int | None = None,
+    num_workers: int = 1,
+    max_num_elements: int = 118,
+    driver: str | None = None,
+    shuffle: bool = True,
+) -> dict[str, LinearReference]:
+    """Fit a set linear references for a list of targets using a given number of batches.
+
+    Args:
+        targets: list of target names
+        dataset: data set to fit linear references with
+        batch_size: size of batch
+        num_batches: number of batches to use in fit. If not given will use all batches
+        num_workers: number of workers to use in data loader
+        max_num_elements: max number of elements in dataset. If not given will use an ambitious value of 118
+        driver: backend used to solve linear system. See torch.linalg.lstsq docs.
+        shuffle: whether to shuffle when loading the dataset
+
+    Returns:
+        dict of fitted LinearReference objects
+    """
+    data_loader = DataLoader(
+        dataset,
+        batch_size=batch_size,
+        shuffle=shuffle,
```

Comment:
For reproducibility maybe dataloader here should take a generator initialized by seed, even if its seed=0? 
I think this shuffle will hit the torch seed, that maybe in a different state depending on what hit it earlier , even for the same original seed.

Reply:
set the seed the same way it is done for the train_loader. Let me know if you had other thoughts!